hearth/lib/hearth/http2/connection.rb

```diff
@@ -0,0 +1,259 @@
+# frozen_string_literal: true
+
+require 'http/2'
+
+module Hearth
+  module HTTP2
+    # An HTTP2 Connection. 1:1 with an open TCP Socket and an http-2 client
+    # # @api private
+    # rubocop:disable Metrics/ClassLength
+    class Connection
+      OPTIONS = {
+        logger: nil,
+        debug_output: false,
+        open_timeout: 15, # in seconds
+        read_timeout: 60,
+        max_concurrent_streams: 100,
+        verify_peer: true,
+        ca_file: nil,
+        ca_path: nil,
+        cert_store: nil,
+        enable_alpn: true,
+        host_resolver: nil
+      }.freeze
+
+      # chunk read size at socket
+      CHUNKSIZE = 1024
+
+      SOCKET_FAMILY = ::Socket::AF_INET
+
+      # 1:1 with an h2 client
+      # 1:1 with a TCP connection
+      # Has a thread that is responsible for reading data to the client
+      def initialize(options = {})
+        OPTIONS.each_pair do |opt_name, default_value|
+          value = options[opt_name].nil? ? default_value : options[opt_name]
+          instance_variable_set("@#{opt_name}", value)
+        end
+
+        @h2_client = ::HTTP2::Client.new
+        @mutex = Mutex.new
+
+        @socket = create_tcp_connection(options)
+        register_h2_callbacks
+        @state = :connected
+        @healthy = true
+
+        @streams = {} # map of stream.id -> stream
+        @errors = []
+        start_connection_thread
+      end
+
+      OPTIONS.each_key do |attr_name|
+        attr_reader(attr_name)
+      end
+
+      alias verify_peer? verify_peer
+
+      def stale?
+        !@healthy || @state != :connected
+      end
+
+      # return a new stream, or nil when max streams is exceeded
+      def new_stream
+        return unless @streams.size < @max_concurrent_streams
+
+        begin
+          stream = @h2_client.new_stream
+          @streams[stream.id] = stream
+          stream
+        rescue ::HTTP2::Error::StreamLimitExceeded
+          nil # exceeded our max streams from remote (server) settings
+        end
+      end
+
+      def close_stream(stream)
+        @mutex.synchronize do
+          @streams.delete(stream.id)
+          stream.close
+        end
+      end
+
+      # all underlying streams will be closed
+      def close
+        @mutex.synchronize do
+          log_debug('closing connection')
+          @status = :closed
+          @healthy = false
+
+          @streams.each_value(&:close)
+          @streams = {}
+          @thread.kill
+
+          @socket&.close
+        end
+      end
+      alias finish close
+
+      private
+
+      # rubocop:disable Metrics
+      def start_connection_thread
```

Comment:
This is an eye sore... is there a better way to reprseent this method?

Reply:
I've done some refactoring (see next commits) - see if that helps improve readability a bit.